tests/net/socket/tls/src/main.c

```diff
@@ -0,0 +1,244 @@
+    zassert_not_equal(r, -1, "failed to set SO_REUSEADDR (%d)", errno);
+
+	r = setsockopt(server_fd, SOL_TLS, TLS_SEC_TAG_LIST,
+		sec_tag_list, sizeof(sec_tag_list));
+	zassert_not_equal(r, -1, "failed to set TLS_SEC_TAG_LIST (%d)", errno);
+
+	memset(&sa, 0, sizeof(sa));
+	r = inet_pton(AF_INET, CONFIG_NET_CONFIG_MY_IPV4_ADDR,
+		&sa.sin_addr.s_addr);
+	sa.sin_family = AF_INET;
+	sa.sin_port = htons(PORT);
+
+	r = bind(server_fd, (struct sockaddr *)&sa, sizeof(sa));
+	zassert_not_equal(r, -1, "failed to bind (%d)", errno);
+
+	r = listen(server_fd, 1);
+	zassert_not_equal(r, -1, "failed to listen (%d)", errno);
+
+    memset(addrstr, '\0', sizeof(addrstr));
+    addrstrp = (char *)inet_ntop(AF_INET, &sa.sin_addr,
+        addrstr, sizeof(addrstr));
+    zassert_not_equal(addrstrp, NULL, "inet_ntop() failed (%d)", errno);
+
+    NET_DBG("listening on [%s]:%d as fd %d",
+        log_strdup(addrstr), ntohs(sa.sin_port), server_fd);
+
+	/* Launch the server thread */
+    NET_DBG("Launching server thread");
+	server_thread_id = k_thread_create(&server_thread, server_stack, STACK_SIZE,
+		server_thread_fn, (void *)(intptr_t)server_fd, NULL, NULL, K_PRIO_PREEMPT(8),
+		0, K_NO_WAIT);
+	zassert_equal(r, 0, "failed to start thread (%d)", r);
+
+	k_thread_name_set(server_thread_id, "server");
+	k_thread_name_set(k_current_get(), "client");
+
+	/* Create client socket */
+	NET_DBG("Creating client socket");
+	r = socket(AF_INET, SOCK_STREAM, IPPROTO_TLS_1_2);
+	zassert_not_equal(r, -1, "failed to create client socket (%d)", errno);
+	client_fd = r;
+
+	r = setsockopt(client_fd, SOL_TLS, TLS_SEC_TAG_LIST,
+		sec_tag_list, sizeof(sec_tag_list));
+	zassert_not_equal(r, -1, "failed to set TLS_SEC_TAG_LIST (%d)", errno);
+
+	NET_DBG("Calling connect()");
+	r = connect(client_fd, (struct sockaddr *)&sa, sizeof(sa));
+	zassert_not_equal(r, -1, "failed to connect (%d)", errno);
+
+	NET_DBG("Calling send()");
+    r = send(client_fd, SECRET, SECRET_SIZE, 0);
+    zassert_not_equal(r, -1, "send() failed (%d)", errno);
+    zassert_equal(SECRET_SIZE, r, "expected: %u actual: %u", SECRET_SIZE, r);
+
+    NET_DBG("Calling recv()");
+    memset(addrstr, 0, sizeof(addrstr));
+    r = recv(client_fd, addrstr, sizeof(addrstr), 0);
+    zassert_not_equal(r, -1, "recv() failed (%d)", errno);
+    zassert_equal(SECRET_SIZE, r, "expected: %u actual: %u", SECRET_SIZE, r);
+
+    zassert_mem_equal(SECRET, addrstr, SECRET_SIZE,
+		"expected: %s actual: %s", SECRET, addrstr);
+
+    NET_DBG("closing client fd");
+    r = close(client_fd);
+    __ASSERT(r != -1, "close() failed on the client fd (%d)", errno);
+
+    NET_DBG("closing server fd");
+    r = close(server_fd);
+    __ASSERT(r != -1, "close() failed on the server fd (%d)", errno);
+}
+
+void test_main(void)
+{
+	k_thread_access_grant(k_current_get(), &server_stack,
+			      &server_thread);
+
+	ztest_test_suite(
+		tls,
+		ztest_unit_test(test_tls)
```

Comment:
I'd consider using some less generic name for the test, in case the testsuite is extended in the future, for instance `test_tls_server_certificate`.

Reply:
Yep. Good call. Maybe the following?
* test_tls_verify_client_none (the usual one-way authentication, client attempting to use cert should fail?)
* test_tls_verify_client_optional (client should be able to connect with + without cert, server requires CA)
* test_tls_verify_client_required (client should not be able to connect without cert, server requires CA)

Reply:
Sounds reasonable.